modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteInMemoryNodeRestartTest.java

```diff
@@ -0,0 +1,354 @@
+    private static final String NODE_BOOTSTRAP_CFG = "{\n"
+            + "  network.port: {},\n"
+            + "  network.nodeFinder.netClusterNodes: {}\n"
+            + "}";
+
+    /** Cluster nodes. */
+    private static final List<Ignite> CLUSTER_NODES = new ArrayList<>();
+
+    private static final List<String> CLUSTER_NODES_NAMES = new ArrayList<>();
+
+    /**
+     * Stops all started nodes.
+     */
+    @AfterEach
+    public void afterEach() throws Exception {
+        var closeables = new ArrayList<AutoCloseable>();
+
+        for (String name : CLUSTER_NODES_NAMES) {
+            if (name != null) {
+                closeables.add(() -> IgnitionManager.stop(name));
+            }
+        }
+
+        CLUSTER_NODES.clear();
+        CLUSTER_NODES_NAMES.clear();
+
+        IgniteUtils.closeAll(closeables);
+    }
+
+    /**
+     * Start node with the given parameters.
+     *
+     * @param idx Node index, is used to stop the node later, see {@link #stopNode(int)}.
+     * @param nodeName Node name.
+     * @param cfgString Configuration string.
+     * @param workDir Working directory.
+     * @return Created node instance.
+     */
+    private static IgniteImpl startNode(int idx, String nodeName, @Nullable String cfgString, Path workDir) {
+        assertTrue(CLUSTER_NODES.size() == idx || CLUSTER_NODES.get(idx) == null);
+
+        CLUSTER_NODES_NAMES.add(idx, nodeName);
+
+        CompletableFuture<Ignite> future = IgnitionManager.start(nodeName, cfgString, workDir.resolve(nodeName));
+
+        if (CLUSTER_NODES.isEmpty()) {
+            IgnitionManager.init(nodeName, List.of(nodeName), "cluster");
+        }
+
+        assertThat(future, willCompleteSuccessfully());
+
+        Ignite ignite = future.join();
+
+        CLUSTER_NODES.add(idx, ignite);
+
+        return (IgniteImpl) ignite;
+    }
+
+    /**
+     * Start node with the given parameters.
+     *
+     * @param testInfo Test info.
+     * @param idx Node index, is used to stop the node later, see {@link #stopNode(int)}.
+     * @return Created node instance.
+     */
+    private IgniteImpl startNode(TestInfo testInfo, int idx) {
+        int port = DEFAULT_NODE_PORT + idx;
+        String nodeName = testNodeName(testInfo, port);
+        String cfgString = configurationString(idx, null, null);
+
+        return startNode(idx, nodeName, cfgString, workDir.resolve(nodeName));
+    }
+
+    /**
+     * Build a configuration string.
+     *
+     * @param idx Node index.
+     * @param cfg Optional configuration string.
+     * @param predefinedPort Predefined port.
+     * @return Configuration string.
+     */
+    private static String configurationString(int idx, @Nullable String cfg, @Nullable Integer predefinedPort) {
+        int port = predefinedPort == null ? DEFAULT_NODE_PORT + idx : predefinedPort;
+        int connectPort = predefinedPort == null ? DEFAULT_NODE_PORT : predefinedPort;
+        String connectAddr = "[\"localhost:" + connectPort + "\"]";
```

Comment:
Practically the same but looks worse imho, because the semicolon should be escaped
`String connectAddr = "[localhost\":\"" + DEFAULT_NODE_PORT + "]";`

Reply:
Now I see, thank you!